Documentation Site (#74)
polymerdao/monomer: doc/docs/learn/create-an-app-with-monomer.md

---
sidebar_position: 2
---

# Create an Application With Monomer

In this tutorial, you will learn how to create a new Monomer application with
a Cosmos SDK [simapp](https://docs.cosmos.network/main/learn/intro/sdk-design#baseapp) (or, `baseapp`).

## Setup

Monomer is a framework for deploying Cosmos SDK applications as rollups on
Ethereum. As such, the development environment for Monomer is similar to that of
a Cosmos SDK application. To begin building an application with Monomer, we can
start with a basic Cosmos SDK app. 

We'll use the handy tool [spawn](https://github.com/rollchains/spawn) to bootstrap a new Cosmos SDK application. If you do not already have it, follow the instructions below to install it.

<details>
  <summary>Install `spawn`</summary>

```bash
git clone https://github.com/rollchains/spawn
cd spawn

> git checkout v0.50.4

Reviewer: Very good idea to checkout a specific version :+1:

make install
```
</details>

### Bootstrapping a Cosmos SDK Application
We can spawn a new Cosmos SDK application with `spawn new`:
```bash
spawn new rollchain --consensus=proof-of-authority \
    --bech32=roll \
    --denom=uroll \
    --bin=rolld \
    --disabled=cosmwasm,globalfee \
    --org=monomer
```

This gives us a boilerplate Cosmos SDK application in the `rollchain`
directory.

## Integrating Monomer
Let's move into the `rollchain` directory and add Monomer to our application.

```bash
cd rollchain

> go get github.com/polymerdao/monomer

Reviewer: Do we need a `go mod tidy` after this? I got some errors when pasting the imports.

Author: Hmm we shouldn't but it might be good to do so. Let me run through this again with a fresh `rollchain` and see if I run into those same errors.

```

### Import `x/rollup` and `x/testmodule`

Now we can import [`x/rollup`](/docs/learn/the-rollup-module) in `app/app.go`. While we're at it, let's also import `x/testmodule` (this
will [initialize a non-empty validator set](https://github.com/polymerdao/monomer/blob/c98eccb49bf857829cadee899359e60fc36e6745/testapp/x/testmodule/module.go#L82) for us so we don't have to do it manually). When you're ready to deploy your application, you can remove `x/testmodule` and configure your own validator set.
```go
import (
    // ...
    "sync"

    "github.com/polymerdao/monomer/x/rollup"
    rolluptypes "github.com/polymerdao/monomer/x/rollup/types"
    rollupkeeper "github.com/polymerdao/monomer/x/rollup/keeper"
    "github.com/polymerdao/monomer/testapp/x/testmodule"

>     testmodulekeeper "github.com/polymerdao/monomer/testapp/x/testmodule"

Reviewer: ```suggestion
    testmodulekeeper "github.com/polymerdao/monomer/testapp/x/testmodule/keeper"
```

Author: Ahh nice catch! Addressed.

    // ...
)
```

> There are a few modifications we need to make to the boilerplate Cosmos SDK app. Namely, we need to initialize the module's store keys, add the module to the genesis module order, and initialize the module in `app/app.go`. Let's walk through these changes.

Reviewer: Better to avoid long paragraphs. 
```suggestion
There are a few modifications we need to make to the boilerplate Cosmos SDK app.

1. Namely, we need to initialize the module's store keys
2. Add the module to the genesis module order
3. Initialize the module in `app/app.go`.

Let's walk through these changes.
```

### Setting the Store Keys
Look for the block that initializes the KV store keys, and add the store key for `x/rollup` and `x/testmodule`:
```go
// app/app.go

func NewChainApp(
  logger log.Logger,
  db dbm.DB,
  traceStore io.Writer,
  loadLatest bool,
  appOpts servertypes.AppOptions,
  baseAppOptions ...func(*baseapp.BaseApp),
) *ChainApp {
    // ...
    keys := storetypes.NewKVStoreKeys(
        // ... 
        ratelimittypes.StoreKey,
        // Monomer
        rolluptypes.StoreKey, // <-- add this line 

>         testmodule.StoreKey,  // <-- add this line

Reviewer: ooooh I like this `<--- add this line`! Can we do the same for the imports above?

        &app.MintKeeper,
        app.BankKeeper,
    ),
),
testmodule.New(testmodulekeeper.New(runtime.NewKVStoreService(keys[testmodule.StoreKey]))),
```

### Finalizing our Changes
Near the bottom of the `NewChainApp` function, comment out the following lines:
```go
// app/app.go

/*protoFiles, err := proto.MergedRegistry()
if err != nil {
    panic(err)
}
err = msgservice.ValidateProtoAnnotations(protoFiles)
if err != nil {
    // Once we switch to using protoreflect-based antehandlers, we might
    // want to panic here instead of logging a warning.
    _, _ = fmt.Fprintln(os.Stderr, err.Error())
}*/
```
and remove this import: `github.com/cosmos/cosmos-sdk/types/msgservice`. 

That's it for `app.go`! Next we'll add a CLI command to interact with Monomer.

### Defining the Monomer Command
In `cmd/rolld/commands.go`, import `github.com/polymerdao/monomer/integrations` and add the following block to the `initRootCmd`
function. This defines the `monomer` command which utilizes our custom start
command hook that handles the integration between Monomer and the Cosmos SDK:
```go
// cmd/rolld/commands.go

import (
    // ...

>     "github.com/polymerdao/monomer/integrations"

Reviewer: ```suggestion
    "github.com/polymerdao/monomer/integrations" // <--- add this line
```

    StartCommandHandler: integrations.StartCommandHandler,
  }))
  rootCmd.AddCommand(monomerCmd)
  // --- End block ---
  
  // ...
}
```

## Building the Application
Before we can build our application, we need to add the following replace directive in `go.mod` and then run `go mod
tidy`: 
```go
github.com/ethereum/go-ethereum => github.com/joshklop/op-geth v0.0.0-20240515205036-e3b990384a74
```

Now we can build our application by running 
```bash
make install 
```
from the root of the
`rollchain/` directory. This will install our chain binary, `rolld`, to the
`$GOBIN` directory.

## Configuring the Application 

There's one last step before we can run our application: We need to write
its genesis file. We can generate one by running
```bash
rolld init my-monomer-app --chain-id=1
```

:::important
Make sure to use a numeric chain ID. Monomer will fail to start if the chain ID
cannot be parsed as a `uint64`.
:::

This will create a new directory in `~/.rollchain/config/` with the necessary
application configuration files. For the sake of this tutorial, we will tweak
the generated `genesis.json` file slightly. Copy the following genesis file to

> `~/.rollchain/config/genesis.json`.

Reviewer: We should add reasoning for tweaking it. How does this differ from the genesis file created on `init` above?

Author: Essentially the only differences between the generated genesis file and our tweaked version is using a numeric value for `chain_id` and including the `testmodule` module. Will add a line to make this clear

Author: Something like this maybe?

---

This will create a new directory in `~/.rollchain/config` with the necessary application configuration files. For the sake of this tutorial, we **must tweak the generated `genesis.json` file** slightly.
1. Update the `chain_id` parameter to a numeric value.
2. Include `x/testmodule` in `app_state` so it can be loaded at genesis.

Copy the following genesis file to `~/.rollchain/config/genesis.json` or make the changes directly.

Author: I pushed the above ^. Lmk if this is good to resolve

Reviewer: Looks good, thanks!
